### PR TITLE
Canvas Medical

### PR DESCRIPTION
**Pull request by radoslawstepinski**
This provides the base for Canvas Medical integration.

What's left:
- We need some good example data to thoroughly test the validation
- We need a sandbox account to test the actual API calls
- In the future it may be a good idea to move the validation to `extensions-core` if we are going to have more FHIR-compliant extensions

New JSON datapoint type: https://github.com/awell-health/extensions-core/pull/6

### CODE REVIEW

File: extensions/healthie/actions/createChartingNote.ts [hunk 1]
```diff
@@ -9,6 +9,13 @@ import { getSdk } from '../gql/sdk'
 import { initialiseClient } from '../graphqlClient'
```
Comment (michal-grzelak) [inline extensions/healthie/actions/createChartingNote.ts:9]: Changes to that file should be removed

File: src/__tests__/extensions.test.ts [hunk 2]
```diff
@@ -23,20 +23,20 @@ describe('Extensions', () => {
   describe('all extension actions have fields labeled correctly', () => {
```
Comment (michal-grzelak) [inline src/__tests__/extensions.test.ts:23]: Changes to that file should be removed

**Review by michal-grzelak — COMMENTED** [review]
Good work 🚀

File: extensions/canvasMedical/actions/getQuestionnaireResponse.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    label: 'Questionnaire Response Id',
+    description: 'The FHIR resource id for the QuestionnaireResponse resource',
+    type: FieldType.STRING,
+    required: true,
+  },
+} satisfies Fields<typeof schemas>
+
+const dataPoints = {
+  questionnaire_response_data: {
+    key: 'questionnaire_response_data',
```
Comment (michal-grzelak) [inline extensions/canvasMedical/actions/getQuestionnaireResponse.ts:27]: Use camel case

File: extensions/canvasMedical/actions/createAppointment.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import { appointmentSchema } from '../validation/dto/appointment.zod'
+import type schemas from '../schemas'
+
+const fields = {
+  appointment_data: {
+    id: 'appointment_data',
+    label: 'Appointment data',
+    description: 'Appointment data',
+    type: FieldType.JSON,
+    jsonType: 'canvas_appointment',
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/createAppointment.ts:23]: Let's remove all the jsonType fields, they're not needed or used in the Awell platform

File: extensions/canvasMedical/actions/createAppointment.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  fields,
+  previewable: true,
+  dataPoints,
+  onActivityCreated: async (payload, onComplete, onError): Promise<void> => {
+    try {
+      const appointment = appointmentSchema.parse(
+        JSON.parse(payload.fields.appointment_data as string)
+      )
+
+      // API Call should produce AuthError or something dif.
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/createAppointment.ts:54]: Not sure I understand what this comment means

File: extensions/canvasMedical/actions/createQuestionnaireResponses.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    description: 'Reference to the Canvas Patient using the patient id',
+    type: FieldType.STRING,
+    required: true,
+  },
+  authored: {
+    id: 'authored',
+    label: 'Authored',
+    description:
+      'Timestamp the Questionnaire response was filled out (If omitted the current timestamp at data ingestion will be used)',
+    type: FieldType.STRING,
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/createQuestionnaireResponses.ts:38]: Should this be FieldType.DATE?

File: extensions/canvasMedical/actions/createQuestionnaireResponses.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+          }
+        : undefined,
+    item,
+    authored,
+    author:
+      authorId !== undefined
+        ? {
+            reference: `Practitioner/${authorId}`,
+          }
+        : undefined,
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/createQuestionnaireResponses.ts:91]: How do we distinguish if the id provided for author is a patient or a practitioner? Here you are assuming it is a practitioner but from the field above it looks like it could be either id?

Either the user should specify somehow which type of id they are providing here, or we should just let the update be done by Canvas bot always

File: extensions/canvasMedical/actions/extractPatientFields.ts
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/extractPatientFields.ts:1]: this will also not be exported since we don't support using a patient object as data point in the Awell Platform yet

File: extensions/canvasMedical/actions/getAppointment.ts
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/getAppointment.ts:1]: This should not be exported as well

File: extensions/canvasMedical/actions/getPatient.ts
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/getPatient.ts:1]: This should not be exported

File: extensions/canvasMedical/actions/getQuestionnaireResponse.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  questionnaireResponseId: {
+    id: 'questionnaireResponseId',
+    label: 'Questionnaire Response Id',
+    description: 'The FHIR resource id for the QuestionnaireResponse resource',
+    type: FieldType.STRING,
+    required: true,
+  },
+} satisfies Fields<typeof schemas>
+
+const dataPoints = {
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/getQuestionnaireResponse.ts:25]: **Note for the future, this action isn't being exported so this isn't feedback on the action itself**

This is an example of the full response object when reading a questionnaire response: 
```
{
  "resourceType": "QuestionnaireResponse",
  "id": "c1ca7b42-b36d-45ef-8e0d-c39f5d27d179",
  "meta": {
    "versionId": "1",
    "lastUpdated": "2021-09-20T13:45:25.229+00:00"
  },
  "questionnaire": "Questionnaire/ac1da1a4-ccc4-492e-a9e0-7f70a58c2129",
  "status": "completed",
  "subject": {
    "reference": "Patient/69dd2c8eb9244ff39c178b3c1f847045"
  },
  "authored": "2021-09-16T19:09:29.530419+00:00",
  "author": {
    "reference": "Practitioner/4150cd20de8a470aa570a852859ac87e"
  },
  "item": [
    {
      "linkId": "2021b11e-b90f-465c-bd16-70b98779272a",
      "text": "Tobacco type",
      "answer": [
        {
          "valueCoding": {
            "system": "http://snomed.info/sct",
            "code": "722496004",
            "display": "Cigarettes"
          }
        },
        {
          "valueCoding": {
            "system": "http://snomed.info/sct",
            "code": "722498003",
            "display": "eCigarette"
          }
        }
      ]
    },
    {
      "linkId": "6359d061-0820-4c36-950c-7a0f996eda98",
      "text": "Tobacco status",
      "answer": [
        {
          "valueCoding": {
            "system": "http://snomed.info/sct",
            "code": "428061000124105",
            "display": "Current Light tobacco user"
          }
        }
      ]
    },
    {
      "linkId": "94651467-f7c5-45d5-b24e-3b2bc42c79c0",
      "text": "Tobacco comment",
      "answer": [
        {
          "valueString": "Would like information about quitting"
        }
      ]
    }
  ]
}
```

https://docs.canvasmedical.com/reference/questionnaireresponse-read

As you can see there are a lot of fields, but most importantly, there is the `item` field which is an array. Without a mechanism to know what is in this array, we won't be able to specify data points to create apart from the static ones in this object (e.g. `status`). If we are not able to extract the answers then the value of this action is very limited

File: extensions/canvasMedical/actions/getQuestionnaireResponse.ts
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/getQuestionnaireResponse.ts:1]: This should not be exported

File: extensions/canvasMedical/actions/getTask.ts
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/getTask.ts:1]: This should not be exported

File: extensions/canvasMedical/actions/index.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+import { extractPatientInfo } from './extractPatientFields'
+
+export const actions = {
+  updatePatient,
+  createPatient,
+  updateAppointment,
+  createAppointment,
+  createTask,
+  extractPatientInfo,
+}
```
[the remark this answers is not in the document]
Comment (rahulkeerthi) [reply extensions/canvasMedical/actions/index.ts:15]: Only setter actions (create, update) should be exported, no getters (read/get), so it's ok to include `create  questionnaire response`  but not the read action for these responses

`extractPatientInfo` should also NOT be exported, since that is only used with the JSON type data points, which we don't support yet

File: extensions/canvasMedical/actions/updateAppointment.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+import { type settings } from '../settings'
+import { makeAPIClient } from '../client'
+import { fromZodError } from 'zod-validation-error'
+import { AxiosError } from 'axios'
+import { appointmentWithIdSchema } from '../validation/dto/appointment.zod'
+import type schemas from '../schemas'
+
+const fields = {
+  appointment_data: {
+    id: 'appointment_data',
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/actions/updateAppointment.ts:19]: camelCase

File: extensions/canvasMedical/client.ts [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+import type {
+  CreatingQuestionnaireResponses,
+  QuestionnaireResponse,
+} from './validation/dto/questionnaireResponses.zod'
+
+interface Id {
+  id: string
+}
+
+interface SerchObject<T> {
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/client.ts:23]: SearchObject?

File: extensions/canvasMedical/index.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+import { type Extension } from '@awell-health/extensions-core'
+import { settings } from './settings'
+import { AuthorType, Category } from '@awell-health/extensions-core'
+import schemas from './schemas'
+
+export const CanvasMedical: Extension<typeof schemas> = {
+  key: 'canvasMedical',
+  title: 'Canvas Medical (BETA)',
+  description:
+    'Canvas Medical is an EHR that is FHIR compliant. NOT FOR PRODUCTION USE',
```
[the remark this answers is not in the document]
Comment (rahulkeerthi) [reply extensions/canvasMedical/index.ts:11]: For now, yes, what we will release is in Beta

File: extensions/canvasMedical/validation/dto/appointment.zod.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    coding: z.array(
+      z.object({
+        system: z.literal('http://snomed.info/sct'),
+        code: z.enum([
+          '439708006',
+          '448337001',
+          '308335008',
+          '31108002',
+          '185317003',
+        ]),
```
[the remark this answers is not in the document]
Comment (rahulkeerthi) [reply extensions/canvasMedical/validation/dto/appointment.zod.ts:59]: @kubiak85 where is this from?

**Review by rahulkeerthi — CHANGES_REQUESTED** [review]
Good stuff @kubiak85!

General
- remove jsonType fields
- use camelCase for ids, etc.
- only export setter actions (update, create)

File: extensions/canvasMedical/actions/getAppointment.ts
Comment (michal-grzelak) [reply extensions/canvasMedical/actions/getAppointment.ts:1]: @rahulkeerthi Just asking, this file is not imported into `index` file in `extensions/canvasMedical/actions/index.ts`. Is there any explanation why this should also NOT be exported from that file? This will break the currently existing tests for that action, because there won't be any export, also probably eslint will show error because there will be file without any export.

File: extensions/canvasMedical/validation/dto/appointment.zod.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (michal-grzelak) [reply extensions/canvasMedical/validation/dto/appointment.zod.ts:59]: @rahulkeerthi I checked the docs and indeed there are some values hardcoded https://docs.canvasmedical.com/reference/appointment-create#appointmenttype, but it is probable that we shouldn't validate it and let Canvas return error when there is no match, because those values may possibly change

File: extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts [hunk 14]
```diff
@@ -4,11 +4,24 @@ import { Category } from '@awell-health/extensions-core'
 import { fromZodError } from 'zod-validation-error'
```
[the remark this answers is not in the document]
Comment (michal-grzelak) [reply extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts:4]: Bump, it should be removed

File: extensions/healthie/actions/createChartingNote.ts (hunk 1 above)
Comment (michal-grzelak) [reply extensions/healthie/actions/createChartingNote.ts:9]: Bump, it should be removed

**Review by michal-grzelak — COMMENTED** [review]
Looks good to me

File: extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (kubiak85) [reply extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts:4]: image

File: extensions/canvasMedical/v1/actions/createTask/config/fields.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
+import { FieldType, type Field } from '@awell-health/extensions-core'
+import { z, type ZodTypeAny } from 'zod'
+
+export const fields = {
+  taskData: {
+    id: 'taskData',
+    label: 'Task data',
+    description: 'Task data',
+    type: FieldType.STRING,
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/v1/actions/createTask/config/fields.ts:9]: Is this a string? Are we not expecting the user to input a JSON object instead?

File: extensions/canvasMedical/v1/actions/createPatient/config/fields.ts [hunk 16]
```diff
@@ -0,0 +1,18 @@
+import { FieldType, type Field } from '@awell-health/extensions-core'
+import { z, type ZodTypeAny } from 'zod'
+
+export const fields = {
+  patientData: {
+    id: 'patientData',
+    label: 'Patient data',
+    description: 'Patient data',
+    type: FieldType.STRING,
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/v1/actions/createPatient/config/fields.ts:9]: @michal-grzelak shouldn't this be JSON?

File: extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts [hunk 17]
```diff
@@ -80,7 +80,7 @@ export const createEmbeddedSignatureRequestWithTemplate: Action<
         userName: signerName,
         clientUserId: patientId,
         returnUrl: replaceStringVariables(returnUrlTemplate, {
-          sessionId,
```
Comment (rahulkeerthi) [inline extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts:83]: is this change relevant in his MR? this might be one that Radoslaw made, but if we can move this to a separate pull request, that would be good

File: src/__tests__/extensions.test.ts (hunk 2 above)
Comment (rahulkeerthi) [reply src/__tests__/extensions.test.ts:23]: @kubiak85 same for this file, if these changes are necessary and not required for this PR then they should go in a separate PR

File: extensions/canvasMedical/v1/actions/index.ts [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+export { createPatient } from './createPatient'
+export { updatePatient } from './updatePatient'
+export { createQuestionnaireResponses } from './createQuestionnaireResponses'
+
+// Note: should also NOT be exported, since that is only used with the JSON type data points, which we don't support yet
+// export { getAppointment } from './getAppointment'
+// export { getTask } from './getTask'
+// export { getPatient } from './getPatient'
+// export { extractPatientInfo } from './extractPatientInfo'
+// export { getQuestionnaireResponse } from './getQuestionnaireResponse'
```
Comment (rahulkeerthi) [inline extensions/canvasMedical/v1/actions/index.ts:14]: okay I think I get the misunderstanding now @kubiak85 -- JSON data points are supported in the Awell platform but what this exactly means was not communicated well. Sorry about that.
See this for explanation: https://awellhealth.slack.com/archives/C04RZELCC1G/p1693564019387499?thread_ts=1693555494.973049&cid=C04RZELCC1G

The FieldType determines two things:
- the default manual input
- the data point types supported in the optional data point selector

When choosing STRING for example, this means the default input is a simple text input, but you can switch to a selector and choose a data point saved in the system (e.g. patient first name) that represents a string (so we filter out dates, numbers, etc).

When choosing a JSON, the same should happen ideally: default is a JSON editor and you can construct an object (which we have today), and the alternative is a Select component and you can choose data points that are storing JSON-parseable content (e.g. objects or arrays). However this second one we don't support today, so when you choose FieldType.JSON we just show the JSON object editor and there is NO option to switch to a selector. This is good, and what we want so you can safely use FieldType.JSON as the type for all the fields where we want the user to construct the JSON object.

The `jsonType` field should still be removed (/not added back in) because that is related to the future support for JSON data points, and is not implemented yet.

**Review by rahulkeerthi — CHANGES_REQUESTED** [review]

File: extensions/canvasMedical/v1/actions/index.ts (hunk 18 above)
Comment (michal-grzelak) [reply extensions/canvasMedical/v1/actions/index.ts:14]: @rahulkeerthi Just FYI, we don't have access to the link you provided (https://awellhealth.slack.com/archives/C04RZELCC1G/p1693564019387499?thread_ts=1693555494.973049&cid=C04RZELCC1G)

But I believe all is clear

**Comment by kubiak85** [comment]
After speaking with @michal-grzelak , I would leave the Radosław code as it is now for other extensions

File: extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts (hunk 17 above)
Comment (rahulkeerthi) [reply extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts:83]: @michal-grzelak can this go in another PR?

File: extensions/canvasMedical/v1/actions/index.ts (hunk 18 above)
Comment (rahulkeerthi) [reply extensions/canvasMedical/v1/actions/index.ts:14]: @michal-grzelak oh that is weird, it was the message from our conversation 
image

File: extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts (hunk 17 above)
Comment (michal-grzelak) [reply extensions/docuSign/actions/createEmbeddedSignatureRequestWithTemplate/createEmbeddedSignatureRequestWithTemplate.ts:83]: What happens in here I believe is that after updating `@awell-health/extensions-core` to new version the type of `sessionId` that is received in action payload changed from `string` to `string | undefined`, whereas this function expects `string`. This means that without this cast, Typescript throws error and tests will fail, so unfortunately from all external changes, this must be included.

**Review by rahulkeerthi — APPROVED** [review]
